File: leetcode_py/Letter Combinations of a Phone Number.py
class Solution:
    def f(self, digits, i, path, res):
        if i == len(digits):
            res.append(path)
            return

        for j in self.maps[ord(digits[i]) - ord('0')]:
            # print path
            self.f(digits, i + 1, path + j, res)

    # @param digits, a string
    # @return a string[]
    def letterCombinations(self, digits):
        if len(digits) == 0:
            return []
        self.maps = ['', '', 'abc', 'def', 'ghi', 'jkl', 'mno', 'qprs', 'tuv', 'wxyz']
        res = []
        self.f(digits, 0, "", res)
        return res


if __name__ == "__main__":
    a = Solution()
    print a.letterCombinations("22")

File: leetcode_js/Letter Combinations of a Phone Number.js
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
	if(digits.length == 0){return [];}
    var maps = ['', '', 'abc', 'def', 'ghi', 'jkl', 
    'mno', 'qprs', 'tuv', 'wxyz'];
    var res = [];
    bfs(res, '', 0);
    return res;

    function bfs(res, path, idx){
    	if(idx == digits.length){
    		res.push(path);
    		return;
    	}
    	var choices = maps[+digits[idx]];
    	for(var c in choices){
    		bfs(res, path+choices[c], idx+1);
    	}
	}
};

